openssl/openssl.go

```diff
@@ -0,0 +1,257 @@
+}
+
+func FreeECGroup(group ECGroup) {
+    C.EC_GROUP_free(group)
+}
+
+func FreeBNCtx(ctx BNCtx) {
+    C.BN_CTX_free(ctx)
+}
+
+func FreeBNMontCtx(montCtx BNMontCtx) {
+    C.BN_MONT_CTX_free(montCtx)
+}
+
+func GetECGroupByCurveNID(curveNid C.int) (ECGroup, error) {
+    // curve must be freed later by the calling function.
+    var curve ECGroup = C.EC_GROUP_new_by_curve_name(curveNid)
+    if curve == nil {
+        // Invalid Curve NID: Curve Group Lookup Failed.
+        return nil, NewOpenSSLError()
+    }
+    return curve, nil
+}
+
+func GetECOrderByGroup(group ECGroup) (BigNum, error) {
+    // order must be freed later by the calling function.
+    var order BigNum = NewBigNum()
+
+    var ctx BNCtx = NewBNCtx()
+    defer FreeBNCtx(ctx)
+
+    result := C.EC_GROUP_get_order(group, order, ctx)
+    if result != 1 {
+        // Invalid Group: Curve Order Lookup Failed.
+        return nil, NewOpenSSLError()
+    }
+    return order, nil
+}
+
+func GetECGeneratorByGroup(group ECGroup) (ECPoint, error) {
+    // generator should not be freed directly by the calling function.
+    // Free the ECGroup instead.
+    var generator ECPoint = C.EC_GROUP_get0_generator(group)
+
+    if generator == nil {
+        // Invalid Group: Generator Lookup Failed.
+        return nil, NewOpenSSLError()
+    }
+    return generator, nil
+}
+
+func GetECGroupDegree(group ECGroup) uint {
+    return uint(C.EC_GROUP_get_degree(group))
+}
+
+func BNIsWithinOrder(checkBN BigNum, curve *Curve) bool {
+    zero, err := IntToBN(0)
+    if err != nil {
+        return false
+    }
+    defer FreeBigNum(zero)
+
+    checkSign := C.BN_cmp(checkBN, zero)
+    rangeCheck := C.BN_cmp(checkBN, curve.Order)
+
+    return checkSign == 1 && rangeCheck == -1
+}
+
+func GetECPointFromAffine(affineX, affineY BigNum, curve *Curve) (ECPoint, error) {
+    // newPoint must be freed later by the calling function.
+    newPoint, err := NewECPoint(curve)
+    if err != nil {
+        return nil, err
+    }
+
+    var ctx BNCtx = NewBNCtx()
+    defer FreeBNCtx(ctx)
+
+    result := C.EC_POINT_set_affine_coordinates_GFp(
+            curve.Group, newPoint, affineX, affineY, ctx)
+    if result != 1 {
+        // Invalid Affine or Curve: EC Point Lookup Failed.
+        return nil, NewOpenSSLError()
+    }
+    return newPoint, nil
+}
+
+func GetAffineCoordsFromECPoint(point ECPoint, curve *Curve) (BigNum, BigNum, error) {
+    // affineX and affineY must be freed later by the calling function.
+    var affineX BigNum = NewBigNum()
+    var affineY BigNum = NewBigNum()
+
+    var ctx BNCtx = NewBNCtx()
+    defer FreeBNCtx(ctx)
+
+    result := C.EC_POINT_get_affine_coordinates_GFp(
+            curve.Group, point, affineX, affineY, ctx)
+    if result != 1 {
+        // Invalid ECPoint or Curve: Affine Lookup Failed.
+        return nil, nil, NewOpenSSLError()
+    }
+    return affineX, affineY, nil
+}
+
+func TmpBNMontCTX(modulus BigNum) (BNMontCtx, error) {
```

Comment:
Probably need to discuss a better pattern for this and see what Go has to offer here. In Python, we can use the contextmanager stuff, but that's not the case with Go. :+1:

Reply:
Would you mind elaborating? I see in pyUmbral that you use contexts with exception handling. What sort of pattern are do you recommend?

Reply:
In Python we can do stuff like:
```
with my_ctx() as ctx:
   blah
```
and `ctx` is collected by the garbage collector when the `with` is exited.

Reply:
`defer` would be the proper way to handle that. I actually just noticed that when the OpenSSL call to `BN_MONT_CTX_set` fails, the memory isn't released. I'll fix this. Other than that it is basically the same as Python's contextmanager except I have to write `defer` or just free it like normal.